File: network.py
#!/usr/bin/env python3

import asyncio
from crypto import *
import keys
import struct

READ_BUFFER_HIGH_WATER_MARK = 16

# 0 - init
# 1 - client handshake
# 2 - server handshake
# 3 - client complete
class TCPProtocol(asyncio.Protocol):
    def __init__(self, wpproto_factory, is_server = False):
        self.is_server = is_server
        self.wpproto = wpproto_factory()
        self.close_exc = None
        self.send_nonce = Nonce(keys.NONCE_SIZE)
        self.recv_nonce = Nonce(keys.NONCE_SIZE)
    def connection_made(self, transport):
        self.transport = transport
        self._paused = False
        self.buffer = b''
        if self.is_server:
            self.handshake = 0
        else:
            # client launch handshake
            challenge = randombytes(32)
            handshake_buf = b'\x01\x03\x02\x43\x82\x43' + keys.pk + challenge
            signature = crypto_sign_detached(handshake_buf, keys.sk)
            self.transport.write(handshake_buf + signature)
            self.handshake = 1
            self.challenge = challenge
        self.wpproto._set_proto(self)
    def connection_lost(self, exc):
        if exc == None:
            exc = self.close_exc
        if self.handshake == 3:
            self.wpproto.connection_lost(exc)
        else:
            self.wpproto.handshake_failed(exc)
    def data_received(self, data):
        self.buffer = self.buffer + data
        if self.handshake == 0:
            # parse client data
            if len(self.buffer) < 6 + keys.PK_SIZE + keys.SIGN_SIZE: # TODO: check the content of the header, calculate the correct length
                return
            device_key = self.buffer[6:6 + keys.PK_SIZE]
            client_challenge = self.buffer[6 + keys.PK_SIZE:6 + keys.PK_SIZE + 32]
            signature = self.buffer[6 + keys.PK_SIZE + 32:6 + keys.PK_SIZE + 32 + keys.SIGN_SIZE]
            handshake_buf = self.buffer[:6 + keys.PK_SIZE + 32]
            try:
                crypto_sign_verify_detached(signature, handshake_buf, device_key)
            except:
                # error verify signature
                self.close_exc = NetworkCryptoException('Invalid signature')
                self.transport.close()
                return
            #self.buffer = self.buffer[6 + keys.PK_SIZE + keys.SIGN_SIZE:]
            self.buffer = b'' # should have no more data
            kex_sk = randombytes(keys.KEX_SK_SIZE)
            kex_pk = crypto_scalarmult_curve25519_base(kex_sk)
            challenge = randombytes(32)
            server_buf = b'\x43\x02\x82' + keys.pk + kex_pk + challenge
            server_sig = crypto_sign_detached(server_buf + client_challenge, keys.sk)
            self.transport.write(server_buf + server_sig)
            self.kex_sk = kex_sk
            self.challenge = challenge
            self.handshake = 2
            self._peer_key = device_key
        elif self.handshake == 1:
            # parse server data
            if len(self.buffer) < 3+keys.PK_SIZE+keys.KEX_PK_SIZE+keys.SIGN_SIZE:
                return
            server_pk = self.buffer[3:3+keys.PK_SIZE]
            kex_pk = self.buffer[3+keys.PK_SIZE:3+keys.PK_SIZE+keys.KEX_PK_SIZE]
            server_challenge = self.buffer[3+keys.PK_SIZE+keys.KEX_PK_SIZE:3+keys.PK_SIZE+keys.KEX_PK_SIZE+32]
            server_sig = self.buffer[3+keys.PK_SIZE+keys.KEX_PK_SIZE+32:3+keys.PK_SIZE+keys.KEX_PK_SIZE+32+keys.SIGN_SIZE]
            signed_buf = self.buffer[:3+keys.PK_SIZE+keys.KEX_PK_SIZE+32]
            self.buffer = b'' # should have no more data
            try:
                crypto_sign_verify_detached(server_sig, signed_buf + self.challenge, server_pk)
            except:
                self.close_exc = NetworkCryptoException('Invalid signature')
                self.transport.close()
                return
            # then complete
            kex_sk = randombytes(keys.KEX_SK_SIZE)
            client_kex_pk = crypto_scalarmult_curve25519_base(kex_sk)
            signature = crypto_sign_detached(client_kex_pk + server_challenge, keys.sk)
            self.transport.write(client_kex_pk + signature)
            shared_secret = crypto_scalarmult_curve25519(kex_sk, kex_pk)
            key = crypto_hash_sha512(shared_secret)
            del self.challenge
            self.recv_key = key[:32]
            self.send_key = key[32:]
            self.handshake = 3
            self._peer_key = server_pk
            # connected
            self.wpproto.connection_made()
        elif self.handshake == 2:
            # parse client data 2
            if len(self.buffer) < keys.KEX_PK_SIZE + keys.SIGN_SIZE:
                return
            client_kex_pk = self.buffer[:keys.KEX_PK_SIZE]
            signature = self.buffer[keys.KEX_PK_SIZE:keys.KEX_PK_SIZE+keys.SIGN_SIZE]
            try:
                crypto_sign_verify_detached(signature, client_kex_pk + self.challenge, self._peer_key)
            except:
                # error verify signature
                self.close_exc = NetworkCryptoException('Invalid signature')
                self.transport.close()
                return
            shared_secret = crypto_scalarmult_curve25519(self.kex_sk, client_kex_pk)
            key = crypto_hash_sha512(shared_secret)
            del self.kex_sk
            del self.challenge
            self.buffer = self.buffer[keys.KEX_PK_SIZE+keys.SIGN_SIZE:]
            self.send_key = key[:32]
            self.recv_key = key[32:]
            self.handshake = 3
            # connected
            self.wpproto.connection_made()
        # may have remain data after handshake
        if self.handshake == 3:
            # unpack data, and send to upper object
            while len(self.buffer) >= 4 + keys.TAG_SIZE:
                length, = struct.unpack('<I', self.buffer[:4])
                if len(self.buffer) < 4 + keys.TAG_SIZE + length:
                    return
                tag = self.buffer[4:4+keys.TAG_SIZE]
                nonce = self.recv_nonce.get()
                ct = self.buffer[4+keys.TAG_SIZE:4+keys.TAG_SIZE+length]
                self.buffer = self.buffer[4+keys.TAG_SIZE+length:]
                try:
                    # sodium combined mode concats mac after ciphertext
                    msg = crypto_aead_chacha20poly1305_ietf_decrypt(ct + tag, b'', nonce, self.recv_key)
                except:
                    # fail to decrypt
                    self.close_exc = NetworkCryptoException('Invalid MAC')
                    self.transport.close()
                    return
                # handle message
                self.wpproto.message_received(msg)
    def pause_writing():
        self.wpproto.pause_writing()
    def resume_writing():
        self.wpproto.resume_writing()
    def write(self, data):
        if self.handshake != 3:
            raise NetworkStateException('handshake not completed')
        # pack data with AEAD, write data
        length = len(data)
        nonce = self.send_nonce.get()
        encrypted = crypto_aead_chacha20poly1305_ietf_encrypt(data, b'', nonce, self.send_key)
        ct, tag = encrypted[:length], encrypted[length:]
        lenbuf = struct.pack('<I', length)
        self.transport.write(lenbuf + tag + ct)
    def close(self):
        self.transport.close()
    def peer_key(self):
        return self._peer_key
    def pause(self):
        if not self._paused:
            self.transport.pause_reading()
            self._paused = True
    def resume(self):
        if self._paused:
            self.transport.resume_reading()
            self._paused = False

class WPProtocol:
    # internal functions
    def _set_proto(self, tcpproto):
        self.tcpproto = tcpproto
    # exported interfaces
    def send_message(self, message):
        self.tcpproto.write(message)
    def peer_key(self):
        return self.tcpproto.peer_key()
    def close(self):
        self.tcpproto.close()
    def pause(self):
        self.tcpproto.pause()
    def resume(self):
        self.tcpproto.resume()
    def drain(self):
        result = asyncio.Future()
        if not hasattr(self, '_drained'):
            self._drained = True
        if self._drained:
            result.set_result(None)
        else:
            if not hasattr(self, '_drainfuture'):
                self._drainfuture = []
            self._drainfuture.append(result)
        return result
    # default implemented flow control
    def pause_writing(self):
        self._drained = False
    def resume_writing(self):
        self._drained = True
        if hasattr(self, '_drainfuture'):
            for f in self._drainfuture:
                f.set_result(None)
        self._drainfuture = []
    # overriden functions
    def handshake_failed(self, exc):
        pass
    def connection_made(self):
        pass
    def connection_lost(self, exc):
        pass
    def message_received(self, message):
        pass

class WPStreamProtocol(WPProtocol):
    def __init__(self, stream_future):
        self.stream_future = stream_future
    def handshake_failed(self, exc):
        self.stream_future.set_exception(exc)
    def connection_made(self):
        stream = WPStream(self)
        self.stream_future.set_result(stream)
        self.stream = stream
    def connection_lost(self, exc):
        self.stream._close(exc)
    def message_received(self, message):
        self.stream._push(message)

class WPListenerProtocol(WPProtocol):
    def __init__(self, conn_handler):
        self.conn_handler = conn_handler
    def handshake_failed(self, exc):
        pass
    def connection_made(self):
        stream = WPStream(self)
        self.stream = stream
        self.conn_handler(stream)
    def connection_lost(self, exc):
        self.stream._close(exc)
    def message_received(self, message):
        self.stream._push(message)

class WPStream:
    def __init__(self, wpproto):
        self.wpproto = wpproto
        self.readbuffer = []
        self.readfuture = []
        self.readable = True
        self.writable = True
        self.exc = None
    def _push(self, data):
        if len(self.readfuture) > 0:
            future = self.readfuture[0]
            del self.readfuture[0]
            future.set_result(data)
        else:
            self.readbuffer.append(data)
            if len(self.readbuffer) >= READ_BUFFER_HIGH_WATER_MARK:
                self.wpproto.pause()
    def _close(self, exc):
        self.readable = False
        self.writable = False
        self.exc = exc
        for future in self.readfuture:
            future.set_exception(NetworkClosedException() if exc == None else exc)
    def read(self):
        result = asyncio.Future()
        if len(self.readbuffer) > 0:
            result.set_result(self.readbuffer[0])
            del self.readbuffer[0]
            if len(self.readbuffer) < READ_BUFFER_HIGH_WATER_MARK:
                self.wpproto.resume()
        else:
            if self.readable:
                self.readfuture.append(result)
                self.wpproto.resume()
            else:
                result.set_exception(NetworkClosedException())
        return asyncio.ensure_future(result)
    def close(self):
        self.writable = False
        self.wpproto.close()
    def write(self, msg):
        if not self.writable:
            raise NetworkClosedException()
        self.wpproto.send_message(msg)
    async def drain(self):
        await self.wpproto.drain()
    def peer_key(self):
        return self.wpproto.peer_key()
    def can_read(self):
        return readable
    def can_write(self):
        return writable
    def get_exception(self):
        return self.exc

class NetworkException(Exception):
    pass
class NetworkCryptoException(NetworkException):
    pass
class NetworkClosedException(NetworkException):
    pass
class NetworkStateException(NetworkException):
    pass

async def connect(addr):
    loop = asyncio.get_event_loop()
    stream_future = asyncio.Future()
    await loop.create_connection(
        lambda: TCPProtocol(lambda: WPStreamProtocol(stream_future)), addr[0], addr[1])
    return await stream_future

async def listen(addr, conn_handler):
    loop = asyncio.get_event_loop()
    server = await loop.create_server(
        lambda: TCPProtocol(lambda: WPListenerProtocol(conn_handler), is_server=True), addr[0], addr[1])
    return server

File: network.js
const co = require('co');
const net = require('net');
const sodium = require('sodium').api;
const crypto = require('crypto');
const chacha = require('chacha');
const fs = require('fs');

const randomBytes = crypto.randomBytes;

let keys = {
    PK_SIZE: sodium.crypto_sign_ed25519_PUBLICKEYBYTES,
    SK_SIZE: sodium.crypto_sign_ed25519_SECRETKEYBYTES,
    SIGN_SIZE: sodium.crypto_sign_BYTES,
    KEY_SIZE: sodium.crypto_aead_chacha20poly1305_ietf_KEYBYTES,
    NONCE_SIZE: sodium.crypto_aead_chacha20poly1305_ietf_NPUBBYTES,
    TAG_SIZE: sodium.crypto_aead_chacha20poly1305_ietf_ABYTES,
    KEX_SK_SIZE: sodium.crypto_scalarmult_curve25519_BYTES,
    KEX_PK_SIZE: sodium.crypto_scalarmult_curve25519_BYTES,
};

try {
    let k = JSON.parse(fs.readFileSync('keys.json').toString());
    keys.pk = new Buffer(k.pk, 'base64');
    keys.sk = new Buffer(k.sk, 'base64');
    keys.type = k.type;
} catch (e) {
    let kp = sodium.crypto_sign_ed25519_keypair();
    let k = {
        pk: kp.publicKey.toString('base64'),
        sk: kp.secretKey.toString('base64'),
        type: 'ed25519'
    }
    fs.writeFileSync('keys.json', JSON.stringify(k));
    keys.pk = kp.publicKey;
    keys.sk = kp.secretKey;
    keys.type = 'ed25519';
}

function Nonce(size) {
    this.nonce = new Buffer(size);
    this.nonce.fill(0);
}
Nonce.prototype.get = function () {
    let result = new Buffer(this.nonce.length);
    this.nonce.copy(result);
    for (let i = 0; i < this.nonce.length; i++) {
        this.nonce[i]++;
        if (this.nonce[i] !== 0) {
            break;
        }
    }
    return result;
}

let connect = co.wrap(function *(host, port) {
    let stream = new Stream(host, port);
    let challenge = randomBytes(32);
    let handshake_buf = Buffer.concat([
        new Buffer('\x01\x03\x02\x43\x82\x43', 'binary'),
        keys.pk,
        challenge
    ]);
    let signature = sodium.crypto_sign_detached(handshake_buf, keys.sk)
    stream.stream.write(handshake_buf);
    stream.stream.write(signature);

    let header = yield stream.readbuf(3);
    let server_pk = yield stream.readbuf(keys.PK_SIZE);
    let server_kex = yield stream.readbuf(keys.KEX_PK_SIZE);
    let server_challenge = yield stream.readbuf(32);
    let server_sig = yield stream.readbuf(keys.SIGN_SIZE);
    let check_buf = Buffer.concat([header, server_pk, server_kex, server_challenge, challenge]);

    if (!sodium.crypto_sign_verify_detached(server_sig, check_buf, server_pk)) {
        throw new Error('invalid signature');
    }

    let kex_sk = randomBytes(keys.KEX_SK_SIZE);
    let kex_pk = sodium.crypto_scalarmult_curve25519_base(kex_sk);
    let signature2 = sodium.crypto_sign_detached(Buffer.concat([kex_pk, server_challenge]), keys.sk);
    stream.stream.write(kex_pk);
    stream.stream.write(signature2);
    let shared_secret = sodium.crypto_scalarmult_curve25519(kex_sk, server_kex);
    let keypair = sodium.crypto_hash_sha512(shared_secret);
    stream.recv_key = keypair.slice(0, 32);
    stream.send_key = keypair.slice(32);
    // handshake finished here

    return stream;
})

function Stream(host, port) {
    this.stream = net.connect(port, host);
    this.readpromise = []
    this.bufdata = new Buffer(0);
    this.send_nonce = new Nonce(keys.NONCE_SIZE); // CHACHA20 Nonce Size = 12!
    this.recv_nonce = new Nonce(keys.NONCE_SIZE);
    this.stream.on('data', (buf) => {
        this.bufdata = Buffer.concat([this.bufdata, buf]);
        this.handlebuf();
    });
    this.stream.on('error', (e) => this.exc = e);
    this.stream.on('close', () => {
        for (let p of this.readpromise) {
            p.reject(new Error('connection closed'));
        }
    });
}
Stream.prototype.readbuf = function (size) {
    let prom = new Promise((resolve, reject) => {
        this.readpromise.push({resolve, reject, size});
    });
    this.handlebuf();
    return prom;
}
Stream.prototype.handlebuf = function () {
    while (this.readpromise.length > 0 && this.bufdata.length >= this.readpromise[0].size) {
        let prom = this.readpromise.splice(0, 1)[0];
        prom.resolve(this.bufdata.slice(0, prom.size));
        this.bufdata = this.bufdata.slice(prom.size);
    }
}

Stream.prototype.write = function (buf) {
    let len = new Buffer(2);
    len.writeUInt16LE(buf.length, 0);
    let cipher = chacha.createCipher(this.send_key, this.send_nonce.get());
    let encbuf = cipher.update(buf);
    cipher.final();
    let tag = cipher.getAuthTag();
    this.stream.write(Buffer.concat([len, tag, encbuf]));
}
Stream.prototype.sendjson = function () {}
Stream.prototype.sendfile = function () {}

Stream.prototype.read = co.wrap(function *() {
    let len = (yield this.readbuf(2)).readUInt16LE(0);
    let tag = yield this.readbuf(keys.TAG_SIZE); // POLY1305 Tag Size = 16!
    let ct = yield this.readbuf(len);
    let decipher = chacha.createDecipher(this.recv_key, this.recv_nonce.get());
    decipher.setAuthTag(tag);
    let data = decipher.update(ct);
    decipher.final();
    return data;
})
Stream.prototype.readjson = function () {}
Stream.prototype.readfile = function () {}

Stream.prototype.end = function () {
    this.stream.end();
}

// test
/*
co(function *() {
    let stream = yield connect('127.0.0.1', 22345);
    stream.write(new Buffer('hello, world'));
    console.log((yield stream.read()).toString());
    stream.write(new Buffer('hello, world'));
    console.log((yield stream.read()).toString());
    stream.write(new Buffer('hello, world'));
    stream.write(new Buffer('hello, world'));
    stream.write(new Buffer('hello, world'));
    console.log((yield stream.read()).toString());
    console.log((yield stream.read()).toString());
    console.log((yield stream.read()).toString());
    stream.end();
})
*/

module.exports = { connect, keys };
